packages/playground/website/src/components/playground-viewport/CodeMirror.tsx

```diff
@@ -0,0 +1,137 @@
+import { EditorView, basicSetup } from 'codemirror';
+import type { ViewUpdate } from '@codemirror/view';
+import { keymap } from '@codemirror/view';
+import { javascript } from '@codemirror/lang-javascript';
+import { json, jsonParseLinter } from '@codemirror/lang-json';
+import { html } from '@codemirror/lang-html';
+import { css } from '@codemirror/lang-css';
+import { markdown } from '@codemirror/lang-markdown';
+import { php } from '@codemirror/lang-php';
+import { linter, lintKeymap, lintGutter } from '@codemirror/lint';
+import { darculaInit } from '@uiw/codemirror-theme-darcula';
+
+import React, {
+  useState,
+  useMemo,
+  useEffect,
+  useRef,
+  useImperativeHandle,
+} from 'react';
+
+export type MemFile = {
+  fileName: string;
+  contents: string;
+};
+
+interface CodeMirrorProps {
+  onChange: (updatedFile: MemFile) => void;
+  initialFile?: MemFile;
+  className?: string;
+}
+
+export type CodeMirrorRef = {
+  setFile: (file: MemFile) => void;
+};
+
+export default React.forwardRef<CodeMirrorRef, CodeMirrorProps>(
+  function CodeMirror({ onChange, initialFile, className = '' }, ref) {
+    const codeMirrorRef = useRef(null);
+    const [file, setFile] = useState<MemFile>(
+      initialFile || {
+        fileName: 'fake.js',
+        contents: "console.log('hello world!');",
+      }
+    );
+
+    useEffect(() => {
+      if (initialFile) {
+        setFile(initialFile)
+      }
+    }, [initialFile]);
+
+    useImperativeHandle(ref, () => ({
+      setFile,
+    }));
+
+    const [counter, setCounter]= useState(0);
+
+    const view = useMemo(() => {
+      if (!codeMirrorRef.current) {
+        setCounter((v) => v+1);
+        return null;
+      }
+
+      const extensions = {
```

Comment:
For posterity: Later on this and https://github.com/WordPress/wordpress-playground/pull/168 may converge into a single `CodeMirror`. For now, let's ship both.